### PR TITLE
br-stream: implement restore command

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31975

Problem Summary:
1. implement restore command for stream.
2. only use table id rewrite rule to perform rewrite.
### What is changed and how it works?
1. add new command br stream restore.
2. collect meta files and data files from stream backup.
3. collect the rewrite rule for each data file.
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/task/stream.go [hunk 1]
```diff
@@ -544,3 +582,150 @@ func RunStreamStatus(
...
+	metas, err := client.ReadStreamMetaByTS(ctx, cfg.RestoreTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(metas) == 0 {
+		log.Info("nothing to restore.")
+		return nil
+	}
+	// read data file by given ts.
+	datas, err := client.ReadStreamDataFiles(ctx, metas, cfg.RestoreTS)
```
Comment (YuJuncen) [inline br/pkg/task/stream.go:674]: handle the error?

File: br/pkg/restore/import.go [hunk 2]
```diff
@@ -626,3 +723,34 @@ func (importer *FileImporter) ingestSSTs(
...
 }
+
+func (importer *FileImporter) downloadAndApplyKVFile(
+	ctx context.Context,
+	file *backuppb.DataFileInfo,
+	rules *RewriteRules,
+	regionInfo *RegionInfo,
+) error {
+	leader := regionInfo.Leader
+	if leader == nil {
```
Comment (YuJuncen) [inline br/pkg/restore/import.go:734]: Maybe backoff until the leader elected?

File: br/pkg/task/stream.go [hunk 3]
```diff
@@ -41,40 +43,56 @@ import (
...
 // StreamConfig specifies the configure about backup stream
 type StreamConfig struct {
 	// common part that all of stream commands need
 	Config
-	TaskName string `json:"task-name" toml:"task-name"`
 
-	// startTs usually equals the tso of full-backup, but user can reset it
+	// FullBackupStorage is used to find the maps between table name and table id during restoration.
+	// if not specified. we cannot apply kv directly.
+	FullBackupStorage string `json:"full-backup-storage" toml:"full-backup-storage"`
```
Comment (YuJuncen) [inline br/pkg/task/stream.go:80]: Maybe move it to `StreamRestoreConfig`?

File: br/cmd/br/stream.go [hunk 4]
```diff
@@ -122,19 +123,47 @@ func newStreamStatusComamnd() *cobra.Command {
...
+		}
+	case task.StreamStart:
+		if err = cfg.ParseStreamStartFromFlags(command.Flags()); err != nil {
+			return errors.Trace(err)
+		}
+		fallthrough
+	default:
+		if err = cfg.ParseStreamCommonFromFlags(command.Flags()); err != nil {
 			return errors.Trace(err)
 		}
```
Comment (YuJuncen) [inline br/cmd/br/stream.go:165]: Maybe remove the `fallthrough` and extract the `PraseStreamCommonFromFlags` out to the `switch`?

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: br/cmd/br/stream.go (hunk 4 above)
Comment (3pointer) [reply br/cmd/br/stream.go:165]: `PraseStreamCommonFromFlags` parse `task-name` config. which is not used is stream restore.🤔. but used for stream start. it cannot be simply move outside.

File: br/pkg/task/stream.go (hunk 3 above)
Comment (3pointer) [reply br/pkg/task/stream.go:80]: currently we don't have `StreamRestoreConfig`. 🤔. I can add a new config. 
`StreamRestoreConfig` and `StreamConfig` both need common `Config` but each has different compatible config.
for example. StreamRestoreConfig needs `full-backup-storage` but doesn't need `task-name`.

so maybe we should change `br stream restore` to `br restore stream`. WDYT?

File: br/pkg/task/stream.go (hunk 1 above)
Comment (3pointer) [reply br/pkg/task/stream.go:674]: Done

File: br/pkg/restore/import.go (hunk 2 above)
Comment (3pointer) [reply br/pkg/restore/import.go:734]: Done

File: br/pkg/restore/client.go [hunk 5]
```diff
@@ -1198,6 +1206,106 @@ func (rc *Client) PreCheckTableClusterIndex(
...
+			m := &backuppb.Metadata{}
+			b, err := rc.storage.ReadFile(ctx, path)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = m.Unmarshal(b)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if m.ReslovedTs < restoreTS {
```
[the remark this answers is not in the document]
Comment (3pointer) [reply br/pkg/restore/client.go]: sure, This PR tried to collect all meta file for now.

**Review by YuJuncen — APPROVED** [review]
LGTM

File: br/cmd/br/stream.go (hunk 4 above)
Comment (3pointer) [reply br/cmd/br/stream.go]: Already add a TODO.

**Comment by YuJuncen** [comment]
/run-integration-br-tests

I have found that with this PR, BR may fail with `ChecksumMismatch`, because rewrite rule not found:

<detail>
[2022/02/09 14:20:12.471 +08:00] [WARN] [import.go:444] ["download file skipped"] [files="{total=1,files=\"[1_124_41_09fa3589bc1c6a7130c4128b9743212d4989bbc28a4f4f2cb8eecbeeb08425e7_1644387040240_write.sst]\",totalKVs=775255,totalBytes=78180884,totalSize=26541360}"] [region="{ID=122,startKey=7480000000000000FF475F728000000000FF34100A0000000000FA,endKey=7480000000000000FF475F728000000000FF3FE60E0000000000FA,epoch=\"conf_ver:1 version:71 \",peers=\"id:123 store_id:1 \"}"] [startKey=7480000000000000FF475F728000000000FF34100A0000000000FA] [endKey=7480000000000000FF475F728000000000FF3FE60E0000000000FA] [file-simple-start=7480000000000000475F72800000000034100A] [file-simple-end=7480000000000000475F7280000000003FE60E] [error="[BR:KV:ErrKVRewriteRuleNotFound]rewrite rule not found"]
</detail>

I'm not sure why... FYI

**Comment by 3pointer** [comment]
> /run-integration-br-tests
> 
> I have found that with this PR, BR may fail with `ChecksumMismatch`, because rewrite rule not found:
> 
> I'm not sure why... FYI

fixed in https://github.com/pingcap/tidb/pull/31821/commits/bd114c53e1a098b24b02b11d5e9a0c34a3ed53ed...

**Comment by 3pointer** [comment]
failed to start tiflash is a known issue.

**Comment by 3pointer** [comment]
> failed to start tiflash is a known issue.

I guess maybe we should update br-stream with master branch to fix br_tiflash case...

File: br/pkg/restore/util.go [hunk 6]
```diff
@@ -307,7 +326,21 @@ func ValidateFileRewriteRule(file *backuppb.File, rewriteRules *RewriteRules) er
...
+		return key, nil
+	}
+	if len(key) > 0 {
+		_, rawKey, _ := codec.DecodeBytes(key, nil)
+		rule := matchOldPrefix(rawKey, rewriteRules)
+		ret := bytes.Replace(rawKey, rule.GetOldKeyPrefix(), rule.GetNewKeyPrefix(), 1)
+		return codec.EncodeBytes([]byte{}, ret), rule
+	}
+	return nil, nil
+}
```
Comment (joccau) [inline br/pkg/restore/util.go:341]: func rewriteEncodedKey(key []byte, rewriteRules *RewriteRules) ([]byte, *import_sstpb.RewriteRule) {
	if rewriteRules == nil  || len(key) == 0 {
		return key, nil
	}
	_, rawKey, _ := codec.DecodeBytes(key, nil)
        return rewriteRawkey(rawKey, rewriteRules)
}

File: br/pkg/restore/util.go [hunk 7]
```diff
@@ -307,7 +326,21 @@ func ValidateFileRewriteRule(file *backuppb.File, rewriteRules *RewriteRules) er
 	return nil
 }
 
-// Rewrites a raw key and returns a encoded key.
+// Rewrites an encoded key and returns a encoded key.
+func rewriteEncodedKey(key []byte, rewriteRules *RewriteRules) ([]byte, *import_sstpb.RewriteRule) {
+	if rewriteRules == nil {
```
Comment (joccau) [inline br/pkg/restore/util.go:331]: func rewriteEncodedKey(key []byte, rewriteRules *RewriteRules) ([]byte, *import_sstpb.RewriteRule) {
if rewriteRules == nil || len(key) == 0 {
return key, nil
}
_, rawKey, _ := codec.DecodeBytes(key, nil)
return rewriteRawkey(rawKey, rewriteRules)
}

Comment (3pointer) [reply br/pkg/restore/util.go:331]: done

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**